### ISSUE TITLE
Sometimes Updater doesn't complete download

### ISSUE THREAD

**Issue by Carlosjrlu** [EnervationRIN/RimworldModUpdater#3]
Sometimes the updater doesn't seem to complete the download, and you get a halfdownloaded mod. For me it happens fairly often.

When this happens, the updater doesn't show the mod in the "Downloaded Mods" tab, so as a workaround I look at that tab before downloading the next mod.

Because the updater knows that the mod wasn't downloaded completely as it doesn't put it in the downloaded tab, a good way to solve this could be to show an alert on the halfdownloaded mods while updating, or maybe retrying until the mod appears in the downloaded tab.

**Comment by EnervationRIN** [comment]
The downloader has no real way of knowing how far alogn a download is or if it has actually failed or not. Since SteamCMD is used for all downloads, you're completely at the mercy of it. I've tried reading standard output and parsing log files to remedy the issue, but nothing really worked properly. I'd recommend looking at the 'logs' folder in SteamCMD to maybe see what the issue is.

To clarify, you say it doesn't appear in the "Downloaded Mods" tab. Does it also disappear from the status bar at the bottom of the "Downloader" tab? ex; "Downloading 1 mods (15MB)".

**Comment by Carlosjrlu** [comment]
Yeah, it appears at the bottom saying "Downloading". And then suddenly "Waiting", after that if I go to the "Downloaded Mods" tab, the mod doesn't appear there.
I just tried to download a mod, failed, looked at the mod folder and it seems it didn't download the textures folder, then tried again and it succeded.
Do you know what log should I be looking for this? I could easily see the difference between this two recent cases.

**Comment by EnervationRIN** [comment]
Try looking at the workshop_log.txt in the 'logs' folder of steamcmd.

Alternatively, delete the steamapps and appcache folder and try again if this is a persistent issue.

**Comment by Carlosjrlu** [comment]
Lol, no difference there.

[2020-11-17 14:31:31] [AppID 294100] Loaded workshop items in "D:\HDD Games\RimWorld\ModUpdater\steamcmd\steamapps\workshop" (0 installed, 0 needed)
[2020-11-17 14:31:31] [AppID 294100] Download item 2007061826 requested by app
[2020-11-17 14:31:32] [AppID 294100] Starting Workshop download job (downloading  )
[2020-11-17 14:31:32] [AppID 294100] Detected workshop change : changed cached item 2007061826, new manifest 8419073435536053243
[2020-11-17 14:31:58] [AppID 294100] Download item 2007061826 result : OK
[2020-11-17 14:31:58] [AppID 294100] Finished Workshop download job : No Error
[2020-11-17 14:31:58] [AppID 294100] Updating workshop item details failed with Failure


[2020-11-17 14:42:21] [AppID 294100] Loaded workshop items in "D:\HDD Games\RimWorld\ModUpdater\steamcmd\steamapps\workshop" (0 installed, 0 needed)
[2020-11-17 14:42:21] [AppID 294100] Download item 2007061826 requested by app
[2020-11-17 14:42:21] [AppID 294100] Starting Workshop download job (downloading  )
[2020-11-17 14:42:21] [AppID 294100] Detected workshop change : changed cached item 2007061826, new manifest 8419073435536053243
[2020-11-17 14:42:31] [AppID 294100] Download item 2007061826 result : OK
[2020-11-17 14:42:31] [AppID 294100] Finished Workshop download job : No Error
[2020-11-17 14:42:31] [AppID 294100] Updating workshop item details failed with Failure

**Comment by EnervationRIN** [comment]
` [2020-11-17 14:31:32] [AppID 294100] Detected workshop change : changed cached item 2007061826, new manifest 8419073435536053243`

This is probably what's causing you trouble; SteamCMD believes it already has it downloaded and "installed" so it doesn't want to download more of it if it's not a new version. The updater should already take care of this for you though. Close the updater and delete the folders `appcache` `depotcache` and `steamapps` and try again.

**Comment by Carlosjrlu** [comment]
Mmm, strange. Did what you said, tried it with 2 never downloaded mods by me, the downloads succeded. But then checked the log and it still says "changed cached item..., new manifest...".

**Comment by EnervationRIN** [comment]
I have no real idea why this happens. I'll try some new things in the next update to prevent it.

**Comment by Carlosjrlu** [comment]
Happens if I try tu update a mod too, in the "Updater" tab.
It won't appear in the "Downloaded Mods" when successful tho, but it will say "Finished updating" at the bottom.

In this example I tried to update one mod with right click, Achtung, it failed, didn't say "Finished updating" and it seemed it got stuck in "updating". Had to restore from backup, close the program and query again.
Tried to download Achtun again, and it worked.
![image](https://user-images.githubusercontent.com/20954701/99559544-e80c0d00-2992-11eb-92a6-e8bb61cc9f9f.png)

I am afraid that I would use the "Update All" button and lose some mods, cause if the mods are half downloaded, they won't be seen in the updater if it doesn't have a "About" folder anymore, nor would it appear in the game.

Am I the only one that has this problem? Had it a few months ago but didn't mind it much cause I had a few mods, but since then I stopped playing rimworld, reinstalled windows 2 times, and redownloaded rimworld for the new version just a few days ago.

**Comment by EnervationRIN** [comment]
I've used the updater plenty and I do not run into this issue anymore. Your best bet would be to delete the updater and re-download it and try again to reset everything.
